File: modules/behaviour/escape.js
const { calculateDistanceBetweenPoints } = require("../../utils/functions");
const mapConfigurations = require("../../utils/mapRegions");

module.exports = class Escape {
  constructor(client, scene, navigation, config, stateManager, settingsManager, stats) {
    this.client = client;
    this.scene = scene;
    this.navigation = navigation;
    this.config = config;
    this.state = stateManager;
    this.settings = settingsManager;
    this.stats = stats;

    this.state.escape.enabled = false;
    this.state.escape.waitedTime = 0;

    this.callback = null;
  }

  start(callback) {
    if (this.state.escape.enabled) return;
    this.state.escape.enabled = true;
    this.activateEscapeLoop();
    this.callback = callback;
  }

  stop() {
    this.state.escape.enabled = false;
    this.callback = null;
    this.state.escape.waitedTime = 0;
  }

  async activateEscapeLoop() {
    while (this.state.escape.enabled) {
      await this.update();
      await delay(100);
    }
  }

  async update() {
    const playerShip = this.scene.getPlayerShip();
    if (!playerShip) return;

    const currentMap = this.scene.currentMap;
    if (!mapConfigurations[currentMap]) {
      console.log(`No portals defined for map: ${currentMap}`);
      return;
    }

    const portals = mapConfigurations[currentMap];
    const closestPortal = this.findClosestPortal(playerShip, portals);

    if (!closestPortal) return;

    // Check if player is being attacked
    const isBeingAttacked = Object.values(this.scene.ships).some((ship) => ship.selected === this.scene.playerId && ship.isAttacking && !ship.name.includes("-=("));

    if (this.scene.safeZone) {
      if (dev) console.log("ESCAPE: In safe zone");

      while (!this.state.detectors.enemy.enemyDetected) {
        await delay(1000);
        this.state.escape.waitedTime += 1000;
        console.log(`Enemy left. Waiting for ${this.state.escape.waitedTime}ms, waited ${this.state.escape.waitedTime / 1000}s`);
        this.stats.messageState = `Escaping - Enemy left. Waiting for ${this.state.escape.waitedTime / 1000}s`;
        if (this.state.escape.waitedTime >= this.settings.escape.delay) {
          if (this.callback) this.callback();
          this.stop();
          return;
        }
      }
      this.state.escape.waitedTime = 0;
      this.stats.messageState = "Escaping - In safe zone, but enemy detected";
    } else if (calculateDistanceBetweenPoints(playerShip.x, playerShip.y, closestPortal.x, closestPortal.y) < 100) {
      if (dev) console.log("ESCAPE: At portal but not in safe zone");
      this.stats.messageState = "Escaping - At portal but not in safe zone";
      if (isBeingAttacked) {
        if (dev) console.log("ESCAPE: Being attacked by player, jumping through portal");
        this.stats.messageState = "Escaping - Being attacked by player, jumping through portal";
        this.client.sendPacket("UserActionsPacket", { actions: [{ actionId: 6 }] });
        await delay(6000);
        this.client.sendPacket("UserActionsPacket", { actions: [{ actionId: 6 }] });
      }
    } else {
      if (dev) console.log(`Escape - Moving to closest portal at (${closestPortal.x}, ${closestPortal.y})`);
      this.config.switchFleeMode();
      await this.navigation.move(closestPortal.x, closestPortal.y);
    }
  }

  findClosestPortal(playerShip, portals) {
    let closestPortal = null;
    let shortestDistance = Infinity;

    for (const portal of portals) {
      // Skip portals leading to PvP zones
      if (portal.to === "T-1" || portal.to === "G-1") continue;

      const distance = calculateDistanceBetweenPoints(playerShip.x, playerShip.y, portal.x, portal.y);

      if (distance < shortestDistance) {
        shortestDistance = distance;
        closestPortal = portal;
      }
    }

    return closestPortal;
  }
};

function delay(ms) {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

File: wu_bot_go/internal/bot/escape.go
package bot

import (
	"context"
	"fmt"
	"strings"
	"time"

	"wu_bot_go/internal/game"
)

// EscapeBehavior handles fleeing from enemy players.
type EscapeBehavior struct {
	scene    *game.Scene
	state    *State
	settings *Settings
	nav      *Navigation
	config   *ConfigSwitcher
	stats    *game.Stats
	sendCh   chan<- game.OutboundPacket
	log      func(string)

	callback func()
}

func NewEscapeBehavior(scene *game.Scene, state *State, settings *Settings, nav *Navigation, config *ConfigSwitcher, stats *game.Stats, sendCh chan<- game.OutboundPacket, log func(string)) *EscapeBehavior {
	return &EscapeBehavior{scene: scene, state: state, settings: settings, nav: nav, config: config, stats: stats, sendCh: sendCh, log: log}
}

// Start begins the escape loop.
func (e *EscapeBehavior) Start(ctx context.Context, callback func()) {
	if e.state.GetEscapeEnabled() {
		return
	}
	e.state.SetEscapeEnabled(true)
	e.state.mu.Lock()
	e.state.Escape.WaitedTime = 0
	e.state.mu.Unlock()
	e.callback = callback

	go e.run(ctx)
}

// Stop stops the escape loop.
func (e *EscapeBehavior) Stop() {
	e.state.SetEscapeEnabled(false)
	e.callback = nil
	e.state.mu.Lock()
	e.state.Escape.WaitedTime = 0
	e.state.mu.Unlock()
}

func (e *EscapeBehavior) run(ctx context.Context) {
	for e.state.GetEscapeEnabled() {
		e.update(ctx)
		select {
		case <-time.After(100 * time.Millisecond):
		case <-ctx.Done():
			return
		}
	}
}

func (e *EscapeBehavior) update(ctx context.Context) {
	ps := e.scene.GetPlayerShip()
	if ps == nil {
		return
	}

	mapName, _, _ := e.scene.GetMapInfo()
	portal := game.FindClosestSafePortal(mapName, ps.X, ps.Y)
	if portal == nil {
		return
	}

	isBeingAttacked := e.checkBeingAttacked()

	if e.scene.GetSafeZone() {
		// In safe zone - wait for enemy to leave
		for !e.state.Detectors.Enemy.EnemyDetected {
			select {
			case <-time.After(1 * time.Second):
			case <-ctx.Done():
				return
			}
			e.state.mu.Lock()
			e.state.Escape.WaitedTime += 1000
			waited := e.state.Escape.WaitedTime
			e.state.mu.Unlock()

			e.stats.SetMessage(fmt.Sprintf("Escaping - Enemy left, waiting %ds", waited/1000))

			if waited >= e.settings.Escape.DelayMs {
				if e.callback != nil {
					e.callback()
				}
				e.Stop()
				return
			}
		}
		e.state.mu.Lock()
		e.state.Escape.WaitedTime = 0
		e.state.mu.Unlock()
		e.stats.SetMessage("Escaping - In safe zone, enemy still present")

	} else if game.Distance(ps.X, ps.Y, portal.X, portal.Y) < 100 {
		e.stats.SetMessage("Escaping - At portal, not safe")
		if isBeingAttacked {
			e.stats.SetMessage("Escaping - Under attack, jumping portal")
			e.sendCh <- game.BuildJumpPortalAction()
			select {
			case <-time.After(6 * time.Second):
			case <-ctx.Done():
				return
			}
			e.sendCh <- game.BuildJumpPortalAction()
		}
	} else {
		e.config.SwitchFleeMode(ctx)
		e.scene.MoveAndWait(ctx, e.sendCh, portal.X, portal.Y)
	}
}

func (e *EscapeBehavior) checkBeingAttacked() bool {
	ships := e.scene.GetShipsSnapshot()
	playerID := e.scene.GetPlayerID()
	for _, ship := range ships {
		if ship.Selected == playerID && ship.IsAttacking && !strings.Contains(ship.Name, "-=(") {
			return true
		}
	}
	return false
}
